presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java

```diff
@@ -336,10 +336,287 @@ public void close()
             throws Exception
     {
         statisticsAggregationOperator.close();
+        systemMemoryContext.setBytes(0);
     }
 
     public interface TableFinisher
     {
         Optional<ConnectorOutputMetadata> finishTable(Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics);
     }
+
+    public interface PageSinkCommitter
+    {
+        ListenableFuture<Void> commitAsync(Collection<Slice> fragments);
+    }
+
+    // A lifespan in a stage defines the unit for commit and recovery in recoverable grouped execution
+    private static class LifespanAndStageStateTracker
+    {
+        private final Map<LifespanAndStage, LifespanAndStageState> noCommitUnrecoverableLifespanAndStageStates = new HashMap<>();
+        private final Map<LifespanAndStage, LifespanAndStageState> taskCommitUnrecoverableLifespanAndStageStates = new HashMap<>();
+
+        // For recoverable execution, it is possible to receive pages of the same lifespan-stage from different tasks. We track all of them and commit the one
+        // which finishes sending pages first.
+        private final Map<LifespanAndStage, Map<TaskId, LifespanAndStageState>> uncommittedRecoverableLifespanAndStageStates = new HashMap<>();
+        private final Map<LifespanAndStage, LifespanAndStageState> committedRecoverableLifespanAndStages = new HashMap<>();
+
+        private final PageSinkCommitter pageSinkCommitter;
+        private final List<ListenableFuture<Void>> commitFutures = new ArrayList<>();
+        private final AtomicLong operatorRetainedMemoryBytes;
+
+        LifespanAndStageStateTracker(
+                PageSinkCommitter pageSinkCommitter,
+                AtomicLong operatorRetainedMemoryBytes)
+        {
+            this.pageSinkCommitter = requireNonNull(pageSinkCommitter, "pageSinkCommitter is null");
+            this.operatorRetainedMemoryBytes = requireNonNull(operatorRetainedMemoryBytes, "operatorRetainedMemoryBytes is null");
+        }
+
+        public void commit()
+        {
+            for (LifespanAndStageState lifespanAndStageState : taskCommitUnrecoverableLifespanAndStageStates.values()) {
+                commitFutures.add(pageSinkCommitter.commitAsync(lifespanAndStageState.getFragments()));
+            }
+
+            ListenableFuture<Void> future = whenAllSucceed(commitFutures).call(() -> null, directExecutor());
+            try {
+                future.get();
+            }
+            catch (InterruptedException e) {
+                future.cancel(true);
+                currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+            catch (ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)